patches/server/0936-Add-Entity-hidden-by-default-flag.patch

```diff
@@ -0,0 +1,207 @@
+index 067d64ce06f626ce90b3ccf608d98f3c64fce335..99ec84ebf018c36ddaf866b1ec05e64cb0f84b43 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+@@ -1721,6 +1721,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+         ChunkMap tracker = ((ServerLevel) this.getHandle().level).getChunkSource().chunkMap;
+         ChunkMap.TrackedEntity entry = tracker.entityMap.get(other.getId());
+         if (entry != null) {
++            entry.showToEvenWhenHiddenByDefault.remove(this.getHandle().connection); // Paper - Entity hiddenByDefault
+             entry.removePlayer(this.getHandle());
+         }
+ 
+@@ -1758,13 +1759,18 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+         if (this.equals(entity)) return;
+ 
+         Set<WeakReference<Plugin>> hidingPlugins = this.hiddenEntities.get(entity.getUniqueId());
+-        if (hidingPlugins == null) {
+-            return; // Entity isn't hidden
+-        }
+-        hidingPlugins.remove(CraftPlayer.getPluginWeakReference(plugin));
+-        if (!hidingPlugins.isEmpty()) {
+-            return; // Some other plugins still want the entity hidden
++        // Paper start - Entity hiddenByDefault
++        if (hidingPlugins != null) {
++            hidingPlugins.remove(CraftPlayer.getPluginWeakReference(plugin));
++            if (!hidingPlugins.isEmpty()) {
++                return; // Some other plugins still want the entity hidden
++            }
++        } else {
++            if (!entity.isHiddenByDefault()) {
++                return; // Entity isn't hidden
++            }
+         }
++        // Paper end
+         this.hiddenEntities.remove(entity.getUniqueId());
+ 
+         // Paper start
+@@ -1783,9 +1789,15 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+         }
+ 
+         ChunkMap.TrackedEntity entry = tracker.entityMap.get(other.getId());
+-        if (entry != null && !entry.seenBy.contains(this.getHandle().connection)) {
+-            entry.updatePlayer(this.getHandle());
++        // Paper start - Entity hiddenByDefault
++        if (entry != null) {
++            entry.showToEvenWhenHiddenByDefault.add(this.getHandle().connection);
++
++            if (!entry.seenBy.contains(this.getHandle().connection)) {
++                entry.updatePlayer(this.getHandle());
++            }
+         }
++        // Paper end
+     }
+     // Paper start
+     private void reregisterPlayer(ServerPlayer player) {
+@@ -1843,7 +1855,17 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+ 
+     @Override
+     public boolean canSee(org.bukkit.entity.Entity entity) {
+-        return !this.hiddenEntities.containsKey(entity.getUniqueId());
++        // Paper start - Entity hiddenByDefault
++        boolean ret = true;
```

Comment:
ret ? might be nice to pick a better variable name here.

Reply:
Absolutely